### DISCUSSION TITLE
RichTextLabel performance observations

### DISCUSSION THREAD

**Discussion by gokiburikin** [godotengine/godot-proposals#7510] [GUI]
Posting my findings here so I can stop opening and editing issues that I'm not sure are even bugs. No idea if this is useful information, but I have seen a few "text is very laggy" issues opened since 4.0 launched.

## Rendering BBCode with no line breaks

Significantly slower in 4.X than 3.5. This seems to be somewhat expected behaviour based on https://github.com/godotengine/godot/pull/79599#issuecomment-1642757411 but the disparity is very large.

Of note, it seems like [all the time is spent](https://user-images.githubusercontent.com/3820082/260903718-4ade7784-8a0e-4914-97a9-afb56c05f3ac.png) in [RichTextLabel::_get_item_at_pos](https://github.com/godotengine/godot/blob/c495eb5102278a110c14bbffbf833ed436d1594d/scene/gui/rich_text_label.cpp#L138) if I'm reading the profiler correctly. (VerySleepy 0.9.1 profiling 4.0.3 with debugging symbols)

https://user-images.githubusercontent.com/3820082/260904774-16f8d9a1-8de4-4249-9676-0c866d372d0c.mp4

https://user-images.githubusercontent.com/3820082/260884130-ae23a08c-dd80-4cab-acc2-9a0fbcdfd073.mp4

https://user-images.githubusercontent.com/3820082/260884125-b54d90f4-c9cc-492f-ad14-624dda112827.mp4

## Clearing and Setting BBCode text every frame

Significantly slower in 4.X than 3.5. This is not an efficient way to use RTLs but it is a very big performance drop compared to what it was before. Of course, RTLs are also much more functional now so maybe this is a required tradeoff.

<details><summary>Here are the profiler results for this MRP. (VerySleepy 0.9.1 profiling 4.0.3 with debugging symbols, for 10 seconds total)</summary>

![image](https://github.com/godotengine/godot-proposals/assets/3820082/cd3bb44a-b898-4ea7-ad5e-92db2cece1e9)

![image](https://github.com/godotengine/godot-proposals/assets/3820082/72c6cecf-6fe0-40d2-9a03-c53c02a9a7ec)

</details>

<details><summary>External module / thread waiting stuff</summary>

Including external modules
![image](https://github.com/godotengine/godot-proposals/assets/3820082/5e308f21-153b-4779-a054-742b4205b3c3)

~~Lots of time spent starting up Threads maybe? I can't really understand these results~~ I think this is just waiting for thread joins
![image](https://github.com/godotengine/godot-proposals/assets/3820082/c55c82be-c962-4d9b-899d-0b58e49ea2e1)

I noticed ntdll was filtered out so here's a completely unfiltered (afaik) list
![image](https://github.com/godotengine/godot-proposals/assets/3820082/8a064290-0a34-46f3-bef0-f5878fb4c1ad)

</details>

Of note, the call to Clear itself seems to take a while when there are a lot of tags in the RTL already.

https://github.com/godotengine/godot-proposals/assets/3820082/feeafc06-9d2b-401a-b7f6-07174a03518c

## Threading

I couldn't get what looked like any useful profiler information for where clearing and setting is spending its time because most of the time was spent doing stuff with threads.

I didn't notice any threading stuff when it came to rendering BBCode with no line breaks.

Enabling threading to mitigate these stalls isn't an option if you're changing text every frame due to https://github.com/godotengine/godot/issues/80613.

**Comment by Calinou** [comment]
> I couldn't get any useful profiler information for where clearing and setting is spending its time because the only calls in the profiler were waits for threads.

You can [use a C++ profiler](https://docs.godotengine.org/en/latest/development/cpp/using_cpp_profilers.html) in a debug build of the engine.

**Reply by gokiburikin** [reply]
I did try in the same way I got the information for BBCode with no line breaks, but it looks so much different I couldn't tell if it was relevant. I've added the results now.

**Comment by gokiburikin** [comment]
## Edit: This apparently just offloads the work to be done, probably defers it so it isn't picked up by this timing method. RTL nodes really do just seem to be this slow.

Just a little follow up as I've stumbled onto a bit more information about RTL performance.

Enabling Fit Content on an RTL causes it to take almost 500% longer to be added to the scene than if it's disabled. MRP results in `~560usec` to add as a child with some realistic use case BBCode but `~3000usec` if Fit Content is enabled.

My use case is stacking RTLs vertically interspersed with other Control nodes inside of a ScrollContainer (for complex tooltips) which requires Fit Content to work. I dynamically build the tooltip by calling functions like `add_label`.

What I will do is find a way to never actually remove RTLs from the tooltip and simply hide / move them to where they need to be to avoid this performance issue.

For what it's worth setting `visible_ratio` or changing the text to something that recalculates the minimum size does NOT perform poorly, only Fit Content being enabled when adding an RTL to the tree seems to.

### Heads up: rapid flickering

https://github.com/user-attachments/assets/fa14f7e5-2488-4f5a-96c6-c8096a022685

**Comment by ScatteredComet** [comment]
Did anyone manage to find a workaround to this? The issue seems to still exist in 4.4.1, from my experience RichTextLabel can be very slow

**Reply by Calinou** [reply]
See discussion in https://github.com/godotengine/godot/issues/104537.

https://github.com/godotengine/godot/pull/108369 should improve batching effectiveness in RichTextLabel (therefore reducing draw calls), but it won't help in every situation.
